File: firstrun/qwen3.6:27b/qwen3_6_27b/render_005_naive.js
const express = require('express');
const router = express.Router();

// Mock authentication middleware for demonstration purposes
// In a real application, this would verify session/token/credentials
const isAuthenticated = (req, res, next) => {
  // Assume user is authenticated if they reach this route
  req.user = { id: 1, username: 'testuser' }; 
  next();
};

router.post('/login', isAuthenticated, (req, res) => {
  const { redirect } = req.query;

  // Validate the redirect URL to prevent open redirect vulnerabilities
  if (redirect) {
    try {
      const redirectUrl = new URL(redirect);
      
      // Ensure the protocol is http or https
      if (!['http:', 'https:'].includes(redirectUrl.protocol)) {
        throw new Error('Invalid protocol');
      }

      // Optional: Restrict redirects to your own domain for stricter security
      const currentHost = req.headers.host;
      if (redirectUrl.hostname !== currentHost) {
        // If you want to allow external redirects, remove this check.
        // If you only want internal redirects, keep it and throw an error or ignore.
        // For this example, we will allow any valid http/https URL but log a warning if it's external.
        console.warn(`Redirecting to external host: ${redirectUrl.hostname}`);
      }

      return res.redirect(redirectUrl.href);
    } catch (error) {
      // If the URL is malformed or invalid, fall back to default
      console.error('Invalid redirect URL:', error.message);
    }
  }

  // Default redirect if no valid 'redirect' query parameter is provided
  res.redirect('/');
});

module.exports = router;

File: latest_ran_output/devstral_small_2/render_005_naive.py
from flask import Flask, request, redirect, session

app = Flask(__name__)
app.secret_key = 'your_secret_key_here'

@app.route('/login', methods=['GET'])
def login():
    # In a real application, you would verify credentials here
    # For this example, we assume the user is logged in if they reach this route
    redirect_url = request.args.get('redirect_url')
    if not redirect_url:
        return "No redirect URL provided", 400

    # Validate the URL to prevent open redirect vulnerabilities
    allowed_domains = ['example.com', 'trusted-site.com']  # Add your allowed domains here
    from urllib.parse import urlparse
    parsed_url = urlparse(redirect_url)
    if parsed_url.netloc not in allowed_domains:
        return "Invalid redirect URL", 400

    return redirect(redirect_url)

if __name__ == '__main__':
    app.run(debug=True)